File: Managed Source/Arsenal.ImageMounter/IO/DisposableDictionary.vb
Imports System.Runtime.Serialization

Namespace IO

    ''' <summary>
    ''' A System.Collections.Generic.Dictionary(Of TKey, TValue) extended with IDisposable implementation that disposes each
    ''' value object in the dictionary when the dictionary is disposed.
    ''' </summary>
    ''' <typeparam name="TKey"></typeparam>
    ''' <typeparam name="TValue"></typeparam>
    <Serializable>
    Public Class DisposableDictionary(Of TKey, TValue As IDisposable)
        Inherits Dictionary(Of TKey, TValue)
        Implements IDisposable

        ' IDisposable
        Protected Overridable Sub Dispose(disposing As Boolean)

            If disposing Then
                '' Dispose each object in list
                For Each value In Values
                    value?.Dispose()
                Next
            End If

            '' Clear list
            Clear()

        End Sub

        ' This code added by Visual Basic to correctly implement the disposable pattern.
        Public Sub Dispose() Implements IDisposable.Dispose
            ' Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub

        Protected Overrides Sub Finalize()
            Dispose(False)
            MyBase.Finalize()
        End Sub

        Public Sub New()
            MyBase.New()

        End Sub

        Public Sub New(capacity As Integer)
            MyBase.New(capacity)

        End Sub

        Public Sub New(dictionary As IDictionary(Of TKey, TValue))
            MyBase.New(dictionary)

        End Sub

        Public Sub New(comparer As IEqualityComparer(Of TKey))
            MyBase.New(comparer)

        End Sub

        Public Sub New(dictionary As IDictionary(Of TKey, TValue), comparer As IEqualityComparer(Of TKey))
            MyBase.New(dictionary, comparer)

        End Sub

        Public Sub New(capacity As Integer, comparer As IEqualityComparer(Of TKey))
            MyBase.New(capacity, comparer)

        End Sub

#If Not NET_CORE Then
        Public Overrides Sub GetObjectData(info As SerializationInfo, context As StreamingContext)
            MyBase.GetObjectData(info, context)
        End Sub

        Protected Sub New(si As SerializationInfo, context As StreamingContext)
            MyBase.New(si, context)
        End Sub
#End If

    End Class

End Namespace

File: Managed Source/Arsenal.ImageMounter/Collections/DisposableDictionary.cs
﻿//  
//  Copyright (c) 2012-2025, Arsenal Consulting, Inc. (d/b/a Arsenal Recon) <http://www.ArsenalRecon.com>
//  This source code and API are available under the terms of the Affero General Public
//  License v3.
// 
//  Please see LICENSE.txt for full license terms, including the availability of
//  proprietary exceptions.
//  Questions, comments, or requests for clarification: http://ArsenalRecon.com/contact/
// 

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;



namespace Arsenal.ImageMounter.Collections;

/// <summary>
/// A System.Collections.Generic.Dictionary(Of TKey, TValue) extended with IDisposable implementation that disposes each
/// value object in the dictionary when the dictionary is disposed.
/// </summary>
/// <typeparam name="TKey"></typeparam>
/// <typeparam name="TValue"></typeparam>
[Serializable]
public class DisposableDictionary<TKey, TValue> : Dictionary<TKey, TValue>, IDisposable
    where TKey : notnull
    where TValue : IDisposable?
{
    // IDisposable
    protected virtual void Dispose(bool disposing)
    {
        if (disposing)
        {
            // Dispose each object in list
            foreach (var value in Values)
            {
                value?.Dispose();
            }
        }

        // Clear list
        Clear();
    }

    // This code added by Visual Basic to correctly implement the disposable pattern.
    public void Dispose()
    {
        // Do not change this code.  Put cleanup code in Dispose(disposing As Boolean) above.
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    ~DisposableDictionary()
    {
        Dispose(false);
    }

    public DisposableDictionary()
        : base()
    {

    }

    public DisposableDictionary(int capacity)
        : base(capacity)
    {

    }

    public DisposableDictionary(IDictionary<TKey, TValue> dictionary)
        : base(dictionary)
    {

    }

    public DisposableDictionary(IEqualityComparer<TKey> comparer)
        : base(comparer)
    {

    }

    public DisposableDictionary(IDictionary<TKey, TValue> dictionary, IEqualityComparer<TKey> comparer)
        : base(dictionary, comparer)
    {

    }

    public DisposableDictionary(int capacity, IEqualityComparer<TKey> comparer)
        : base(capacity, comparer)
    {

    }

#if NET8_0 || NET8_0_OR_GREATER
    [Obsolete("This API supports obsolete formatter-based serialization. It should not be called or extended by application code.", DiagnosticId = "SYSLIB0051", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
#endif
    public override void GetObjectData(SerializationInfo info, StreamingContext context)
        => base.GetObjectData(info, context);

#if NET5_0_OR_GREATER
    [Obsolete("This API supports obsolete formatter-based serialization. It should not be called or extended by application code.", DiagnosticId = "SYSLIB0051", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
#endif
    protected DisposableDictionary(SerializationInfo si, StreamingContext context)
        : base(si, context)
    {
    }
}